File: ascii_fields/animations/burning_ship.py
from ..core import Animation, clamp, render_field
from ._fractal import iteration_cap, mandelbrot_grid, zoom_scale


class BurningShipAnimation(Animation):
  """Endless zoom into the Burning Ship fractal -- the jagged, flame-like
  cousin of the Mandelbrot set, formed by taking absolute values each step."""

  thresholds = [
    (0.04, " "), (0.13, "."), (0.25, ":"), (0.38, "-"), (0.52, "="),
    (0.66, "+"), (0.78, "*"), (0.90, "#"), (1.01, "@"),
  ]

  CENTER = (-1.7549, -0.0260)

  def render(self, width, height, elapsed, phase, options):
    scale, _ = zoom_scale(elapsed * 0.6, base=0.9, depth=9.0, speed=0.5)
    max_iter = iteration_cap(width, height)
    cx, cy = self.CENTER
    grid = mandelbrot_grid(width, height, cx, cy, scale, max_iter, ship=True)
    contrast = options.contrast
    return render_field(width, height,
                        ([clamp(v * contrast) for v in row] for row in grid),
                        options, self)

File: src/animations/burning_ship.rs
use crate::animation::{Animation, FrameContext};
use crate::core::{clamp, render_field, FieldStyle};
use super::fractal_base::{axes, iteration_cap, mandelbrot_cell};

const STYLE: FieldStyle = FieldStyle { gray_lo: 234, gray_hi: 255, default_theme: "fire" };
const TH: &[(f64, char)] = &[
  (0.04, ' '), (0.13, '.'), (0.25, ':'), (0.38, '-'), (0.52, '='),
  (0.66, '+'), (0.78, '*'), (0.90, '#'), (1.01, '@'),
];
const CENTER: (f64, f64) = (-1.7549, -0.0260);

pub struct BurningShip;

impl Animation for BurningShip {
  fn render(&mut self, ctx: &FrameContext, out: &mut String) {
    let log_z = 9.0 * 0.5 * (1.0 - (ctx.elapsed * 0.6 * 0.5).cos());
    let scale = 0.9 * (-log_z).exp();
    let max_iter = iteration_cap(ctx.width, ctx.height, 60, 110);
    let (xs, ys) = axes(ctx.width, ctx.height, CENTER.0, CENTER.1, scale);
    let contrast = ctx.options.contrast;
    let mut grid = vec![0.0_f64; ctx.width * ctx.height];
    for (row, &ci) in ys.iter().enumerate() {
      let base = row * ctx.width;
      for (col, &cr) in xs.iter().enumerate() {
        grid[base + col] = clamp(mandelbrot_cell(cr, ci, max_iter, true) * contrast);
      }
    }
    render_field(ctx, &grid, TH, &STYLE, out);
  }
}
